ktor-server/ktor-server-netty/jvm/src/io/ktor/server/netty/cio/NettyResponsePipeline.kt

```diff
@@ -6,138 +6,94 @@ package io.ktor.server.netty.cio
-
-    private suspend fun fill() {
-        tryFill()
-        if (running.isEmpty()) {
-            return fillSuspend()
+    /**
+     * Flush if there is some unflushed data, nothing to read from channel and no active requests
+     */
+    internal fun flushIfNeeded() {
+        if (isWriteOccurred.get() && isChannelReadComplete.get() && activeRequests.get() == 0L) {
+            context.flush()
+            isWriteOccurred.set(false)
         }
     }
 
-    private fun tryFill() {
-        while (isNotFull()) {
-            if (!pollReady()) {
-                tryStart()
-                dst.read()
-                break
-            }
+    internal fun processResponse(call: NettyApplicationCall) {
+        call.previousCallFinished = previousCallHandled
+        call.callFinished = context.newPromise()
+        previousCallHandled = call.callFinished
 
-            tryStart()
-        }
+        processElement(call)
     }
 
-    private suspend fun fillSuspend() {
-        if (running.isEmpty()) {
-            @OptIn(ExperimentalCoroutinesApi::class)
-            val e = incoming.receiveCatching().getOrNull()
+    private fun processElement(call: NettyApplicationCall) {
+        val previousCallFinished = call.previousCallFinished
 
-            if (e != null && e.ensureRunning()) {
-                running.addLast(e)
-                tryFill()
-            }
-        }
-    }
-
-    private fun pollReady(): Boolean {
-        for (index in 1..(readyQueueSize - ready.size)) {
-            val e = incoming.tryReceive().getOrNull() ?: return false
-            ready.addLast(e)
+        if (previousCallFinished.isDone && !previousCallFinished.isSuccess) {
+            processCallFailed(call, previousCallFinished.cause())
+            return
         }
-        return true
-    }
 
-    private fun tryStart() {
-        while (ready.isNotEmpty() && running.size < runningQueueSize) {
-            val e = ready.removeFirst()
-            if (e.ensureRunning()) {
-                running.addLast(e)
-            } else {
-                break
+        call.response.responseFlag.addListener {
+            if (!it.isSuccess) {
+                processCallFailed(call, it.cause())
+                return@addListener
             }
-        }
-    }
 
-    private fun isNotFull(): Boolean = ready.size < readyQueueSize || running.size < runningQueueSize
-
-    private fun hasNextResponseMessage(): Boolean {
-        tryFill()
-        return running.peekFirst()?.isCompleted == true
-    }
-
-    @Suppress("NOTHING_TO_INLINE")
-    private suspend inline fun processElement(element: NettyRequestQueue.CallElement) {
-        val call = element.call
-
-        try {
-            processCall(call)
-        } catch (actualException: Throwable) {
-            processCallFailed(call, actualException)
-        } finally {
-            call.responseWriteJob.cancel()
+            call.previousCallFinished.addListener waitPreviousCall@{ previousCallResult ->
```

Comment:
I would really like to see a brief comment here in a following manner:
```
// We need to wait until the previous call is finished and only proceed if it was successfully processed.
// See more: <link to some helpful information source>
```

Reply:
Moved to the separate function and added comment